apps/content-gateway-api/src/app/repository/MongoUserRepository.ts

```diff
@@ -0,0 +1,163 @@
+import { createLogger } from "@banklessdao/util-misc";
+import {
+    ContentGatewayUser,
+    DatabaseError,
+    UserCreationError,
+    UserDeletionError,
+    UserNotFoundError,
+    UserRepository,
+    UserUpdateError
+} from "@domain/feature-gateway";
+import { pipe } from "fp-ts/lib/function";
+import * as TE from "fp-ts/TaskEither";
+import { Db, ObjectId } from "mongodb";
+import { wrapDbOperation } from ".";
+import { MongoUser } from "./mongo/MongoUser";
+
+export const mongoUserToCGUser = (user: MongoUser): ContentGatewayUser => {
+    return {
+        id: user._id.toString(),
+        name: user.name,
+        roles: user.roles,
+        apiKeys: user.apiKeys,
+    };
+};
+
+type Deps = {
+    db: Db;
+    collName: string;
+};
+
+export const createMongoUserRepository = async ({
+    db,
+    collName,
+}: Deps): Promise<UserRepository> => {
+    const logger = createLogger("MongoUserRepository");
+    const users = db.collection<MongoUser>(collName);
+    await users.createIndex({ name: 1 });
+    await users.createIndex({ roles: 1 });
+    await users.createIndex({ "apiKeys.id": 1 });
+    await users.createIndex({ "apiKeys.hash": 1 });
+
+    const findById = (
+        id: string
+    ): TE.TaskEither<UserNotFoundError | DatabaseError, ContentGatewayUser> => {
+        return pipe(
+            wrapDbOperation(() => users.findOne({ _id: new ObjectId(id) }))(),
+            TE.chainW((mongoUser) => {
+                if (mongoUser) {
+                    return TE.right(mongoUserToCGUser(mongoUser));
+                } else {
+                    return TE.left(
+                        new UserNotFoundError(`Couldn't find user by id ${id}`)
+                    );
+                }
+            })
+        );
+    };
+
+    const findByApiKeyId = (
```

Comment:
Is there a db-access each time when some API function is called, or is the key/user cached somehow?

Reply:
MongoDB does this automatically. We haven't reached the point when this becomes a problem, (although we already have the indexing problem that @theobayard is solving.